### PR TITLE
Reference counted uncanonicalised overlay

### PR DESCRIPTION
**Pull request by arkpar**
This removed linear performance dependency on the number of uncanonicalised blocks for a bit of memory tradeoff. Helps with performance now that we have uncanonicalised blocks limit increased to 4096.

### CODE REVIEW

File: core/state-db/src/noncanonical.rs [hunk 1]
```diff
@@ -112,9 +116,35 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
...
+	fn insert_values(values: &mut HashMap<Key, (u32, DBValue)>, inserted: Vec<(Key, DBValue)>) {
+		for (k, v) in inserted.into_iter() {
+			debug_assert!(values.get(&k).map_or(true, |(_, value)| *value == v));
+			let (ref mut counter, _) = values.entry(k).or_insert_with(|| (0, v));
+			*counter += 1;
+		}
+	}
+
+	fn discard_values(values: &mut HashMap<Key, (u32, DBValue)>, inserted: Vec<Key>) {
+		for k in inserted.into_iter() {
```
Comment (rphmeier) [inline core/state-db/src/noncanonical.rs]: redundant `into_iter`

File: core/state-db/src/noncanonical.rs [hunk 2]
```diff
@@ -112,9 +116,35 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 			parents,
 			pending_canonicalizations: Default::default(),
 			pending_insertions: Default::default(),
+			values: values,
 		})
 	}
 
+	fn insert_values(values: &mut HashMap<Key, (u32, DBValue)>, inserted: Vec<(Key, DBValue)>) {
+		for (k, v) in inserted.into_iter() {
```
Comment (rphmeier) [inline core/state-db/src/noncanonical.rs]: here too

File: core/state-db/src/noncanonical.rs [hunk 3]
```diff
@@ -329,7 +362,8 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 					level.last().expect("Hash is added in `insert` in reverse order").hash == hash)
 				.expect("Hash is added in insert");
 
-			self.levels[level_index].pop();
+			let	overlay = self.levels[level_index].pop().expect("Empty levels are not allowed in self.leves");
```
Comment (rphmeier) [inline core/state-db/src/noncanonical.rs]: typo `self.levels`

File: core/state-db/src/noncanonical.rs [hunk 4]
```diff
@@ -112,9 +116,35 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 			parents,
 			pending_canonicalizations: Default::default(),
 			pending_insertions: Default::default(),
+			values: values,
 		})
 	}
 
+	fn insert_values(values: &mut HashMap<Key, (u32, DBValue)>, inserted: Vec<(Key, DBValue)>) {
```
Comment (rphmeier) [inline core/state-db/src/noncanonical.rs]: free functions are more idiomatic

**Comment by rphmeier** [comment]
looks good except for grumbles

File: core/state-db/src/noncanonical.rs (hunk 4 above)
Comment (arkpar) [reply core/state-db/src/noncanonical.rs]: Not sure why. Member function have a benefit of using generic types that the struct is parameterized with, without having to specify them again along with the trait bounds, etc.

**Pull request merged**